src/org/rust/lang/core/resolve/util/RustResolveUtil.kt

```diff
@@ -10,22 +13,49 @@ public object RustResolveUtil {
         while (current != null) {
             when (current) {
                 is RustResolveScope -> return current
-                else -> current = current.parent
+                else                -> current = current.parent
             }
         }
 
         return null
     }
 
-    fun getGlobalResolveScopeFor(elem: PsiElement): RustResolveScope {
-        var current = getResolveScopeFor(elem)
-        while (true)
-        {
-            val p = current?.let { getResolveScopeFor(it) } ?: break
-            current = p;
+    fun getCrateRootFor(elem: PsiElement): RustResolveScope? {
+        val mod = getSelfModFor(elem) ?: return null
+
+        val superMod = getSuperModFor(mod)
+        return if (superMod == null) mod else getCrateRootFor(superMod)
+    }
+
+    fun getSuperModFor(elem: RustModItem): RustModItem? {
```

Comment:
It's also under-estimated in its correctness: things like `#[path = ]` and nested module decls make this style of resolution impossible =)

I think the proper way to handle this would be to maintain a tree of modules for each crate. Such tree can be constructed correctly if we traverse modules once in a top down fashion from the crate root. We should extract knowledge about crate roots from cargo and maybe even have some sort of an action like `mark file as a crate root`.